server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java

```diff
@@ -678,11 +633,26 @@ private static boolean waitingShardsStartedOrUnassigned(SnapshotsInProgress snap
     }
 
     private static boolean removedNodesCleanupNeeded(SnapshotsInProgress snapshotsInProgress, List<DiscoveryNode> removedNodes) {
-        // If at least one shard was running on a removed node - we need to fail it
-        return removedNodes.isEmpty() == false && snapshotsInProgress.entries().stream().flatMap(snapshot ->
-                StreamSupport.stream(((Iterable<ShardSnapshotStatus>) () -> snapshot.shards().valuesIt()).spliterator(), false)
```

Comment:
This was really evil for large snapshots (even more so in the context of concurrent snapshots). We would iterate all the shards even if no nodes were removed or the snapshot finished already.